Song qc sidecar#1

KevinFHartmann opened the pull request.

KevinFHartmann committed:
Change: Renamed handler.py to verifier.py
Feature: Defined and implemented a verification protocol for the server

The server reads in a JSON string to verify, and returns a JSON string reply.
The JSON consists of an object with two fields: status, and details.

When status is 'OK', it means that the verifier ran successfully, and that details is a list of strings. Each string is an error message describing an issue that the verifier found with the json provided in the message. If there were no issue, details is empty list.

When status is 'BAD_JSON', it means that the message wasn't a JSON string, and details is an error string from the JSON parser.

When status is 'VERIFIER_ERROR', it means that the verifier didn't return a list of strings ー either it threw an exception, or it returned some python object that wasn't a list of strings. In either case, details is an error string describing the error message.

KevinFHartmann committed:
Bugfix: We should pass in the original json message after we ensure that
it can be deserialized as JSON, so that the verifier can deserialize it to
any set of python objects that it prefers.

KevinFHartmann committed:
Bugfix: Fixed double print in run.py
Bugfix: Set content type to JSON in Dockerfile
Change: Wrote a more plausible verifier

KevinFHartmann committed:
Bugfix: test_run.py wasn't getting json anymore.

KevinFHartmann committed:
Change: Added an ISSUES status distinct from OK status, to make it clearer on the SONG server end.
Change: Simplified the return type to always return a list of details (easier to model as a DTO in SONG)

KevinFHartmann committed:
Bugfix: BAD_JSON is another client input issues to fix. It shouldn't have it's own status code.

rtisma commented on Dockerfile:
```diff
@@ -18,7 +19,7 @@ WORKDIR /home/app/
 RUN chown -R app:app ./ &&   chmod -R 777 /home/app/python
 USER app
-ENV fprocess="python3 run.py"
+ENV fprocess="python3 run.py" content_type="application/json"
```
The user `app` must have the uid 9999 and gid 9999, and the USER must be $APP_UID (not $APP_USER)


So add this
```
ENV APP_USER app
ENV APP_UID 9999
ENV APP_GID 9999
RUN addgroup -S -g $APP_GID $APP_USER  \
    && adduser -S -u $APP_UID -G $APP_USER $APP_USER

USER $APP_UID
```

rtisma replied:
i think this should be moved up higher, to line 8

rtisma commented on tests/test_run.py:
```diff
@@ -0,0 +1,50 @@
+                   "list contains non-strings"),
+                  ('{"name": "Rosi"}',
+                   '{"status": "VERIFIER_ERROR", "details": ["Verifier exception: UnknownPerson(\'Rosi\')"]}',
+                   "Verifier throws exception")]
```
This is hard to read and maintain. Please put this in a json file and move it into a tests/fixtures (or similar) directory

rtisma commented on verifier.py:
```diff
@@ -0,0 +1,55 @@
+    "sampleType":null,"specimen":{"submitterSpecimenId":"abcd"},"donor":{"donorId":null,"studyId":null,"gender":null, 
+    "submitterDonorId":"DO1234"}}],"files":[{"fileName":"test.bam.gz","fileType":"BAM"},{"fileName":"test.fasta",
+    "fileType":"FASTA"}]} '''
+    print(verify(analysis))
```
These tests should not be here, they should be in a test directory. This code should only be for verification functionality

rtisma reviewed (changes requested):
Some small changes. In addition, we want the bioinfo testing to be easy. Ideally, they would simply add a json payload to the `tests/passing` directory, and that file would automatically be used during the tests to verify its passing. Like wise with the `tests/failing` directory. This means they should only really be touching:
- the verifyer.py:verify function
- the `tests/passing` directory
- the `tests/failing` directory

KevinFHartmann committed:
Fixes as per PR

KevinFHartmann committed:
Enhancement: Added test for verifier

KevinFHartmann committed:
Enhancement: Wrote example json test files for the sample verifier
Enhancement: Fixed bugs in the test_verifier and improved it's reporting (doesn't die after first error)

rtisma reviewed (changes requested):
docker container does not build

KevinFHartmann committed:
Bugfix: Re-arranged permission related code so that it would be executed under the right permissions, removed redundant code so that the app user wasn't created or specified twice

rtisma commented on run.py:
```diff
@@ -1,22 +1,56 @@
+
+def protocol(message, verify_function=verifier.verify):
+    try:
+        json.loads(message)
```
handle the case where message == None. Currently, when this method i called via GET request with no body, i get 
`"BAD JSON: Expecting value: line 1 column 1 (char 0)"`
would be better to say
`"BAD JSON: null json"`
or something similar, indicating to the user, they sent null

rtisma commented on tests/failing/bad_json:
```diff
@@ -0,0 +1 @@
+<json>This isn't json</json>
```
for all the `failing/*` and `passing/*` files, please append `.json`. This way, when the file is loaded in any editor, it can highlight syntax properly for json

rtisma reviewed (changes requested):
Please update the README.md with instructions on how to run the tests

KevinFHartmann committed:
Bugfix: Put the tests path in the pytests so that pytests can be run from the command line root directory.
Enhancement: Added description of usage, including how to use tests, in the README

rtisma commented on README.md:
````diff
@@ -1,2 +1,34 @@
+
+```
+docker build . -t sidecar
+docker run -p8080:8080 sidecar
````
typo, should be
docker run -p 8080:8080  sidecar

rtisma reviewed (changes requested):
- a small typo in the readme
- test files still need to be suffixes with .json
- null handling still needs to be done for verify method

KevinFHartmann committed:
Bugfix: Typo in README
Bugfix: Nicer reporting on empty message
Bugfix: Renamed test inputs to end with ".json"

KevinFHartmann committed:
Bugfix: In run.py, Python uses "not", not !
Bugfix: In run.py, mixed tabs & spaces.
Code Cleanup: In verifier.py, Python convention is snake case, not camel case

rtisma commented on tests/failing/no_bai_file.json:
```diff
@@ -0,0 +1,5 @@
+{"analysisType": {"name": "variantCall"},
+ "files": [{"fileType": "BAM", "fileName": "sample1.bam", "fileMd5sum":"aabbcc"},
+           {"fileType": "CRAM", "fileName":"sample2.cram", fileMd5sum:"a1b2c3"}
```
typo here. should be `"fileMd5sum"` with quotes

rtisma commented on tests/failing/two_files.json:
```diff
@@ -0,0 +1,5 @@
+{"analysisType": {"name": "variantCall"},
+ "files": [{"fileType": "BAM", "fileName": "sample1.bam", "fileMd5sum":"aabbcc"},
+           {"fileType": "CRAM", "fileName":"sample2.cram", fileMd5sum:"a1b2c3"}
```
same here

rtisma commented on README.md:
````diff
@@ -1,2 +1,34 @@
+To run the tests, run 
+
+```
+python3 -m pytest
````
although this runs the test, having a bit verbosity may help the developer ensure their fixture ran. change this to
`python3 -m pytest -s -v`. 

The output
```
Testing file no_bai_file.json (should fail)
FAIL (good -- failure expected)
```
is super useful!

rtisma reviewed (changes requested):
minor changes

KevinFHartmann committed:
Bugfix: Added flags -s -v to pytest
Code Cleanup: Fixed python warnings and improper JSON files.
